### ISSUE TITLE
More info on disabling peers

### ISSUE THREAD

**Issue by cmcaine** [tonarino/innernet#9]
Hi, this looks like useful software, thank you for releasing it open source!

In the readme you say:

> ### Enabling/Disabling Peers
>
> For security reasons, IP addresses cannot be re-used by new peers, and therefore peers cannot be deleted. However, they can be disabled. Disabled peers will not show up in the list of peers when fetching the config for an interface.

Could you expand on the security properties here? Specifically:

1. When will other peers start to reject connections from a disabled peer?
2. How are other peers told to reject these connections?

Thanks!

**Comment by mcginty** [comment]
Thanks @cmcaine, and thanks for opening the issue for clarification. Sorry for the delay in responding.

> Could you expand on the security properties here? Specifically:
> 1. When will other peers start to reject connections from a disabled peer?
> 2. How are other peers told to reject these connections?

Sure! Once a peer is disabled, the server will stop sending that peer in the `/user/state` call that all clients make to get the visible CIDRs and peers. When updating, clients will remove peers from their interface that are no longer reported from the server. So after disabling a peer, the rejection from peers is largely based on how frequently those peers fetch updates from the server. The default client systemd unit fetches every 60 seconds, for example.

Note that while the client won't have the peer on its interface, it won't delete the peer data from its cache (https://github.com/tonarino/innernet/blob/main/client/src/data_store.rs#L63) so the server can't late report the peer as re-enabled but with a different public key.

**Comment by KhalilSantana** [comment]
Would you consider adding an option to force-remove a peer?  I basically want to utilize innernet as a base overlay network/vpn to provide services across nodes, however I need to be able to restructure stuff without having to destroy the entire network. As it stands, if the user ever fat-fingers or changes his mind on his addressing, the only options would be:

1) Clone/move the old innernet settings from under it and hope it works from device a -> b (also pretty unsafe, as the old peer obviously had access to the old keys)
2) Do hacky things like directly editing conf files and dbs
3) Give up abd have a messy network

I'd also love a `recreate-peer`which would re-key the peer and ask the same questions as add-peer with the previous values as defaults.

**Comment by mcginty** [comment]
@KhalilSantana yeah, I'm definitely in favor of adding a rekey feature, because you're right that at the moment the only way to do that feels hacky/messy (creating a new peer and disabling the old one).

It's not perfectly clear exactly the cleanest implementation for it yet in my head though. Two options would be:
1. `innernet rekey` would basically do `disable` -> `new peer` -> `rename`, thus preserving the simple `(IP, pubkey)` pinning that peers currently do to mitigate attacks from a compromised coordinating server.
2. `innernet rekey` would somehow convince clients to replace their `(IP, pubkey_old)` pin with `(IP, pubkey_new)`, possibly by having the peers sign messages attesting to the new key being announced. This is more complicate because there are currently no keys compatible with signature algorithms, without either
  a.  [converting curve25519 keys to ed25519 keys](https://moderncrypto.org/mail-archive/curves/2014/000205.html) which would be a sensitive operation that should be verified to be safe, or
  b. generating new signing keys and storing them alongside the WireGuard curve25519 keys.

**Comment by KhalilSantana** [comment]
I suggest two different functions/subcommands:

* `innernet rekey` which would re-gen the peer keys/invite but keep it's current IP address.
* `innernet recreate` which would ask all the same questions of `innernet new` but defaulting to the current values instead of `new`'s behavior. This would allow changing anything about a peer, including keeping the same (IP, key) tuple, or changing it.

Use cases:
* "I want to tidy up my network addressing" -> `innernet recreate`
* "A peer might have it's keys compromised`-> `innernet rekey`
* "User forgot to redeem invite before the deadline" -> `innernet recreate` (Maybe this would also address #90)

> mitigate attacks from a compromised coordinating server.

That would be the innernet-server host? If so, I don't really consider that a particularly big downside, or rephrasing, I consider that the usability benefits  outweighs security downsides related to that.
